### ISSUE TITLE
Expert-informed feature bagging

### ISSUE THREAD

**Issue by dgrnbrg** [microsoft/LightGBM#3375]
## Summary

I would like to control feature bagging rather than having it be random. I'd like to specify, for a given iteration, the subset of features that I want that tree to be fit to.

## Motivation

For my use of lightgbm, I've found that my models underfit when using random feature bagging, but if I manually select subsets of features that make sense together, I'm able to achieve much higher performance than random feature bagging or using all the candidate features.

## Description

I would like to either have a callback on each iteration to specify the features to use for that tree, or a the ability to specify a set of feature subsets that can be randomly selected from each iteration.

**Comment by btrotta** [comment]
We have recently added the ability to exclude specific feature interactions: https://lightgbm.readthedocs.io/en/latest/Parameters.html#interaction_constraints
It sounds like you want the opposite feature to this (ability to specify the _allowed_ feature interactions), but it might be helpful nevertheless (depending on your specific use case).

**Comment by dgrnbrg** [comment]
Ah, so supposing that I had 2 correlated conditioning features (A & B), and a predictive feature (P), then I could ensure that any path in any tree has (A&P or B&P), but never (A&B or A&B&P)? Is there a reference where I can read more about this?

While that could be helpful for my application, I actually would like something like the learning rate callback, but for feature selection for a given tree. There are many combinations of interactions I'd like, and I'd like to have the model learn (randomly or deterministically) from only the feature subsets that I choose.

Could I generate interaction constraints to do that w/o hitting pathological algorithmic performance?

**Comment by btrotta** [comment]
@dgrnbrg See my edit to my earlier comment above, I had misremembered how this works. Our implementation is similar to xgboost's; see their docs for detailed explanation: https://xgboost.readthedocs.io/en/latest/tutorials/feature_interaction_constraint.html

For your example, assuming your features are in the order [A, B, P], you could specify `interaction_constraints=[[A,P], [B,P]]` to prevent A and B appearing in the same branch.
